### PR TITLE
test(api): e2e for iam provider on rds models

### PR DESCRIPTION
**Pull request by sundersc**
Adding E2E tests for Auth Directive IAM provider support.
- Covers IAM public and private rules. (Public rule strictly gives access to unauthRule and private rule gives access to the logged in users).
- Mutli-auth scenarios with API key, lambda and IAM for queries, mutations and subscriptions.

### CODE REVIEW

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts [hunk 1]
```diff
@@ -95,7 +95,7 @@ const convertAuthRoleToVtl = (
           type: str(role.strategy),
           provider: str('iam'),
           roleArn: role.strategy === 'public' ? ref('ctx.stash.unauthRole') : ref('ctx.stash.authRole'),
-          cognitoIdentityPoolId: hasIdentityPoolId ? ref('ctx.stash.identityPoolId') : nul(),
+          ...(role.strategy === 'private' && { cognitoIdentityPoolId: identityPoolId ? str(identityPoolId) : nul() }),
```
Comment (alharris-at) [inline packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts:98]: You need to reference the identityPoolId in stash, right? I don't think that value `identityPoolId` you're referencing exists. This should fail to build if I'm not mistaken.

Comment (alharris-at) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts:98]: Yeah, just double-checked, and it looks like the build is failing on this.

File: packages/amplify-e2e-core/src/init/amplifyPush.ts [hunk 2]
```diff
@@ -43,12 +45,18 @@ export function amplifyPush(cwd: string, testingWithLatestCodebase = false): Pro
...
+    const pushCommands = spawn(getCLIPath(testingWithLatestCodebase), ['push'], { cwd, stripColors: true, noOutputTimeout: pushTimeoutMS })
       .wait('Are you sure you want to continue?')
-      .sendConfirmYes()
-      .wait('Do you want to generate code for your newly created GraphQL API')
-      .sendConfirmNo()
-      .wait(/.*/)
+      .sendConfirmYes();
+
+    if (!settings?.skipCodegen) {
+      pushCommands.wait('Do you want to generate code for your newly created GraphQL API')
```
Comment (phani-srikar) [inline packages/amplify-e2e-core/src/init/amplifyPush.ts:53]: There is an existing method already for this called `amplifyPushWithoutCodegen` that's used in other tests.

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts (hunk 1 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts:98]: Ah, looks rebase didn't go well. Looking into it.

File: packages/amplify-e2e-tests/src/utils/api.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import { API } from 'aws-amplify';
+
+export const reconfigureAmplifyAPI = (url: string, region: string, appSyncAuthType: string, apiKey?: string): void => {
+  if (appSyncAuthType === 'API_KEY') {
```
[the remark this answers is not in the document]
Comment (sundersc) [reply packages/amplify-e2e-tests/src/utils/api.ts:4]: No strong preference here. Like the more explicit check on the auth type here.

File: packages/amplify-e2e-core/src/init/amplifyPush.ts (hunk 2 above)
Comment (sundersc) [reply packages/amplify-e2e-core/src/init/amplifyPush.ts:53]: Not sure why we have a separate method for these minor differences. Still I feel like this is a good addition as we can eventually get rid of the `amplifyPushWithoutCodegen` with this change.

File: packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts (hunk 1 above)
Comment (sundersc) [reply packages/amplify-graphql-auth-transformer/src/vtl-generator/rds/resolvers/common.ts]: PR has been updated now with the correct changes.

File: packages/amplify-e2e-core/src/init/amplifyPush.ts (hunk 2 above)
Comment (phani-srikar) [reply packages/amplify-e2e-core/src/init/amplifyPush.ts]: Can you just add a task for this to migrate the usage of `amplifyPushWithoutCodegen`? Don't really want two methods that do the same thing.

**Review by alharris-at — APPROVED** [review]

File: packages/amplify-e2e-tests/src/__tests__/rds-mysql-auth-apikey-lambda.test.ts [hunk 4]
```diff
@@ -27,7 +27,7 @@ describe('RDS Relational Directives', () => {
   // Generate settings for RDS instance
   const username = db_user;
   const password = db_password;
-  let region = 'ap-northeast-2';
+  let region = 'us-east-1';
```
Comment (phani-srikar) [inline packages/amplify-e2e-tests/src/__tests__/rds-mysql-auth-apikey-lambda.test.ts:30]: We're talking about allow listing our E2E account to test this in ICN first. Can keep this the same.

**Comment by phani-srikar** [comment]
The build job is still failing

**Comment by sundersc** [comment]
All the tests are passing now.

**Review by alharris-at — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Pull request merged**